File: services/gateway/src/domain/pipeline.ts
import { config } from '../config';
import { id } from '../logger';
import type { AuditLog } from './audit';
import { evaluate } from './careGate';
import { classify, verticalFor } from './contentClass';
import type { Graph } from './graph';
import { finaliseSurface, intendedSurface, widgetFamilyFor } from './momentRouter';
import { checkTone, stripSensitiveFinancials } from './tone';
import type { Moment, RawEvent, Reason, Surface } from './types';

interface DayCounter {
  day: string;
  pushes: number;
}

/**
 * The single path from a raw event to a surface. Order is fixed and total:
 *   classify → route(intended) → tone lint → Care Gate → finalise surface →
 *   strip financials → audit → emit.
 * A blocked event is still audited (with reasons) but never emitted.
 */
export class Pipeline {
  private alertCounters = new Map<string, DayCounter>();

  constructor(
    private readonly graph: Graph,
    private readonly audit: AuditLog,
    private readonly emit: (userId: string, moment: Moment) => void,
  ) {}

  process(event: RawEvent, now = new Date()): Moment | null {
    const user = this.graph.getUser(event.userId);
    if (!user) return null;

    const contentClass = classify(event.kind);
    const vertical = verticalFor(event.kind);
    const intended = intendedSurface(event.kind);

    // Reject dark-pattern copy up front — it can never become a moment.
    const tone = checkTone(event.title, event.body);
    if (!tone.ok) {
      return this.blocked(user.id, event, contentClass, [
        {
          code: 'CONTENT_DARK_PATTERN',
          message: `Rejected urgency/pressure language: "${tone.matched}".`,
        },
      ], now);
    }

    const boostCapReached =
      contentClass === 'inducement' ? this.graph.isBoostCapReached(user.id, event) : false;

    const decision = evaluate(user, contentClass, {
      now,
      intendedSurface: intended,
      dailyAlertCapReached: this.isPushCapReached(user.id, now),
      boostCapReached,
    });

    if (!decision.allowed) {
      return this.blocked(user.id, event, contentClass, decision.reasons, now);
    }

    const surface = finaliseSurface(event.kind, decision);
    if (surface === 'push') this.incrementPush(user.id, now);
    if (contentClass === 'inducement') this.graph.recordBoostDelivery(user.id, event);

    // No spend/loss ever reaches an ambient surface.
    const { data, stripped } = stripSensitiveFinancials(event.data, surface);
    const reasons: Reason[] = [...decision.reasons];
    if (stripped) {
      reasons.push({
        code: 'FINANCIALS_STRIPPED',
        message: 'Spend/loss data removed — ambient surfaces show status only.',
      });
    }

    let cta = event.cta;
    if (contentClass === 'informational' && cta && event.kind !== 'cashout_window') {
      cta = undefined;
      reasons.push({
        code: 'CTA_STRIPPED',
        message: 'Informational continuity only — promotional CTA removed.',
      });
    }

    const widgetFamily = surface === 'widget' ? widgetFamilyFor(event.kind) : undefined;

    const moment: Moment = {
      id: id('mom_'),
      userId: user.id,
      eventId: event.id,
      kind: event.kind,
      vertical,
      contentClass,
      surface,
      widgetFamily,
      cta,
      title: event.title,
      body: event.body,
      data,
      reasons,
      createdAt: now.toISOString(),
    };

    this.audit.record({
      id: id('aud_'),
      userId: user.id,
      eventId: event.id,
      kind: event.kind,
      vertical,
      contentClass,
      allowed: true,
      surface,
      widgetFamily,
      reasons,
      createdAt: now.toISOString(),
    });

    this.emit(user.id, moment);
    return moment;
  }

  private blocked(
    userId: string,
    event: RawEvent,
    contentClass: ReturnType<typeof classify>,
    reasons: Reason[],
    now: Date,
  ): null {
    this.audit.record({
      id: id('aud_'),
      userId,
      eventId: event.id,
      kind: event.kind,
      vertical: verticalFor(event.kind),
      contentClass,
      allowed: false,
      reasons,
      createdAt: now.toISOString(),
    });
    return null;
  }

  // --- push (alert) cap bookkeeping ---
  private dayKey(now: Date): string {
    return now.toISOString().slice(0, 10);
  }
  private counter(userId: string, now: Date): DayCounter {
    const today = this.dayKey(now);
    const existing = this.alertCounters.get(userId);
    if (existing && existing.day === today) return existing;
    const fresh: DayCounter = { day: today, pushes: 0 };
    this.alertCounters.set(userId, fresh);
    return fresh;
  }
  private isPushCapReached(userId: string, now: Date): boolean {
    return this.counter(userId, now).pushes >= config.DEFAULT_DAILY_ALERT_CAP;
  }
  private incrementPush(userId: string, now: Date): void {
    this.counter(userId, now).pushes += 1;
  }
}

File: services/gateway-rs/src/domain/pipeline.rs
use chrono::{DateTime, Utc};
use uuid::Uuid;

use crate::domain::care_gate::evaluate;
use crate::domain::content_class::{classify, widget_family_for};
use crate::domain::moment_router::{finalise_surface, intended_surface};
use crate::domain::tone::{check_tone, strip_sensitive_financials};
use crate::domain::types::{
    AuditRecord, CareGateContext, ContentClass, EventKind, Moment, RawEvent, Reason, ReasonCode,
    Surface,
};
use crate::state::AppState;

fn id(prefix: &str) -> String {
    format!("{}{}", prefix, Uuid::new_v4().simple())
}

fn now_iso(now: DateTime<Utc>) -> String {
    now.to_rfc3339_opts(chrono::SecondsFormat::Millis, true)
}

fn offer_key_and_cap(state: &AppState, event: &RawEvent) -> (String, u32) {
    if event.kind == EventKind::BoostOffer {
        let store = state.store.read().expect("store read");
        if let Some(user) = store.get_user(&event.user_id) {
            if let Some(boost) = user.boosts.first() {
                return (boost.id.clone(), boost.weekly_cap);
            }
        }
    }
    (format!("{:?}", event.kind), 5)
}

pub fn process(state: &AppState, event: RawEvent, now: DateTime<Utc>) -> Option<Moment> {
    let user = {
        let store = state.store.read().expect("store read");
        store.get_user(&event.user_id).cloned()
    }?;

    let content_class = classify(event.kind);
    let vertical = event.kind.vertical();
    let intended = intended_surface(event.kind);

    let tone = check_tone(&event.title, &event.body);
    if !tone.ok {
        let reasons = vec![Reason::new(
            ReasonCode::ContentDarkPattern,
            format!("Rejected urgency/pressure language: {}", tone.matched.unwrap_or_default()),
        )];
        record_blocked(state, &event, content_class, reasons, now);
        return None;
    }

    let (offer_key, offer_cap) = offer_key_and_cap(state, &event);
    let offer_cap_reached = if content_class == ContentClass::Inducement {
        let store = state.store.read().expect("store read");
        store.is_offer_cap_reached(&event.user_id, &offer_key, offer_cap)
    } else {
        false
    };

    let ctx = CareGateContext {
        intended_surface: intended,
        hourly_alert_cap_reached: state.hourly_cap_reached(&user.id, now),
        daily_alert_cap_reached: state.daily_cap_reached(&user.id, now),
        offer_cap_reached,
    };

    let decision = evaluate(&user, content_class, &ctx, now);
    if !decision.allowed {
        record_blocked(state, &event, content_class, decision.reasons, now);
        return None;
    }

    let surface = finalise_surface(event.kind, &decision);
    if surface == Surface::Push {
        state.increment_push(&user.id, now);
    }
    if content_class == ContentClass::Inducement {
        let mut store = state.store.write().expect("store write");
        store.record_offer_delivery(&user.id, &offer_key);
    }

    let (data, stripped) = strip_sensitive_financials(&event.data);
    let mut reasons = decision.reasons.clone();
    if stripped {
        reasons.push(Reason::new(
            ReasonCode::FinancialsStripped,
            "Spend/loss data removed. Ambient surfaces show status only.",
        ));
    }

    let mut cta = event.cta.clone();
    if content_class == ContentClass::Informational
        && cta.is_some()
        && event.kind != EventKind::CashoutWindow
    {
        cta = None;
        reasons.push(Reason::new(
            ReasonCode::CtaStripped,
            "Informational continuity only. Promotional CTA removed.",
        ));
    }

    let widget_family = if surface == Surface::Widget {
        widget_family_for(event.kind)
    } else {
        None
    };

    let created_at = now_iso(now);
    let moment = Moment {
        id: id("mom_"),
        user_id: user.id.clone(),
        event_id: event.id.clone(),
        vertical,
        kind: event.kind,
        content_class,
        surface,
        widget_family,
        cta,
        title: event.title.clone(),
        body: event.body.clone(),
        data,
        reasons: reasons.clone(),
        created_at: created_at.clone(),
    };

    state.record_audit(AuditRecord {
        id: id("aud_"),
        user_id: user.id.clone(),
        event_id: event.id.clone(),
        vertical,
        kind: event.kind,
        content_class,
        allowed: true,
        surface: Some(surface),
        widget_family,
        reasons,
        created_at,
    });

    state.hub.publish(&user.id, moment.clone());
    Some(moment)
}

fn record_blocked(
    state: &AppState,
    event: &RawEvent,
    content_class: ContentClass,
    reasons: Vec<Reason>,
    now: DateTime<Utc>,
) {
    state.record_audit(AuditRecord {
        id: id("aud_"),
        user_id: event.user_id.clone(),
        event_id: event.id.clone(),
        vertical: event.kind.vertical(),
        kind: event.kind,
        content_class,
        allowed: false,
        surface: None,
        widget_family: None,
        reasons,
        created_at: now_iso(now),
    });
}
